idea/src/org/jetbrains/kotlin/idea/intentions/AddJvmOverloadsIntention.kt

```diff
@@ -23,38 +23,57 @@ import org.jetbrains.kotlin.idea.util.addAnnotation
 import org.jetbrains.kotlin.idea.util.findAnnotation
 import org.jetbrains.kotlin.name.FqName
 import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.endOffset
+import org.jetbrains.kotlin.psi.psiUtil.startOffset
 
 private val annotationFqName = FqName("kotlin.jvm.JvmOverloads")
 
-class AddJvmOverloadsIntention : SelfTargetingIntention<KtParameterList>(
-        KtParameterList::class.java, "Add '@JvmOverloads' annotation"
+class AddJvmOverloadsIntention : SelfTargetingIntention<KtModifierListOwner>(
+        KtModifierListOwner::class.java, "Add '@JvmOverloads' annotation"
 ), LowPriorityAction {
 
-    override fun isApplicableTo(element: KtParameterList, caretOffset: Int): Boolean {
-        val parent = element.parent as? KtModifierListOwner ?: return false
-        val target = when (parent) {
-            is KtNamedFunction -> "function '${parent.name}'"
-            is KtPrimaryConstructor -> "primary constructor"
-            is KtSecondaryConstructor -> "secondary constructor"
+    override fun isApplicableTo(element: KtModifierListOwner, caretOffset: Int): Boolean {
+        val (targetName, parameterList) = when (element) {
+            is KtNamedFunction -> {
+                val funKeyword = element.funKeyword ?: return false
+                val valueParameterList = element.valueParameterList ?: return false
+                if (caretOffset !in funKeyword.startOffset..valueParameterList.endOffset) {
+                    return false
+                }
+
+                "function '${element.name}'" to valueParameterList
+            }
+            is KtSecondaryConstructor -> {
+                val constructorKeyword = element.getConstructorKeyword()
+                val valueParameterList = element.valueParameterList ?: return false
+                if (caretOffset !in constructorKeyword.startOffset..valueParameterList.endOffset) {
+                    return false
+                }
+
+                "secondary constructor" to valueParameterList
+            }
+            is KtPrimaryConstructor -> "primary constructor" to element.valueParameterList
             else -> return false
         }
-        text = "Add '@JvmOverloads' annotation to $target"
+        val parameters = parameterList?.parameters ?: return false
+
+        text = "Add '@JvmOverloads' annotation to $targetName"
 
         return !ProjectStructureUtil.isJsKotlinModule(element.getContainingKtFile())
-               && element.parameters.any { it.hasDefaultValue() }
-               && parent.findAnnotation(annotationFqName) == null
+               && parameters.size > 1
```

Comment:
This is not exactly the check that I meant. Specifically for the primary constructor, Kotlin will generate a no-argument overload when all arguments have default values, even if the @JvmOverloads annotation is not present. This doesn't happen for regular functions.